plugin/pkg/admission/alwayspullimages/admission.go

```diff
@@ -66,6 +67,28 @@ func (a *AlwaysPullImages) Admit(ctx context.Context, attributes admission.Attri
 		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
 	}
 
+	if attributes.GetOperation() == admission.Update {
+		oldPod, ok := attributes.GetOldObject().(*api.Pod)
+		if !ok {
+			return apierrors.NewBadRequest("Resource was marked with kind Pod but old object was unable to be converted")
+		}
+
+		oldImages := sets.NewString()
+		pods.VisitContainersWithPath(&oldPod.Spec, field.NewPath("spec"), func(c *api.Container, _ *field.Path) bool {
+			oldImages.Insert(c.Image)
+			return true
+		})
+		newImages := sets.NewString()
+		pods.VisitContainersWithPath(&pod.Spec, field.NewPath("spec"), func(c *api.Container, _ *field.Path) bool {
+			newImages.Insert(c.Image)
```

Comment:
I guess we can check and short-circuit here as soon as we see a new image, rather than accumulating the full set of new images:
```
hasNewImage := false
pods.VisitContainersWithPath(&pod.Spec, field.NewPath("spec"), func(c *api.Container, _ *field.Path) bool {
  if !oldImages.Has(c.Image) {
    hasNewImage = true
  }
  // keep checking additional containers as long as we haven't seen a new image yet
  return !hasNewImage
})
```

Reply:
thanks, nice improvement.

Reply:
the ut finished in 0.6s on my mac now, better than 0.9s before the change.